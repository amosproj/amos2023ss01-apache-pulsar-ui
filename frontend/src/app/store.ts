// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2010-2021 Dirk Riehle <dirk@riehle.org
// SPDX-FileCopyrightText: 2019 Georg Schwarz <georg. schwarz@fau.de>

import { configureStore } from '@reduxjs/toolkit'
import counterReducer from '../features/counter/counterSlice'

export const store = configureStore({
	reducer: {
		counter: counterReducer,
	},
})

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch
