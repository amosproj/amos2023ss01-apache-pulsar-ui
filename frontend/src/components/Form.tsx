import React, { useState, useEffect } from 'react'
import { TextField, Button } from '@mui/material'
import TopicSelect from './CustomSelect'

const Form = () => {
	const [topic, setTopic] = useState<string>('')
	const [message, setMessage] = useState<string>('')
	const [topicError, setTopicError] = useState<boolean>(false)
	const [messageError, setMessageError] = useState<boolean>(false)
	const [data, setData] = useState<Array<any>>([])

	const getData = () => {
		fetch('data/dummy.json', {
			headers: {
				'Content-Type': 'application/json',
				Accept: 'application/json',
			},
		})
			.then(function (response) {
				console.log(response)
				return response.json()
			})
			.then(function (myJson) {
				console.log(myJson)
				setData(myJson)
			})
	}

	useEffect(() => {
		getData()
	}, [])

	const handleSubmit = (event: any) => {
		event.preventDefault()

		setMessageError(false)
		setTopicError(false)

		if (message == '') {
			setMessageError(true)
		}
		if (topic == '') {
			setTopicError(true)
		}

		if (message && topic) {
			console.log(message, topic)
		}

		const dataCopy = [...data]

		dataCopy.map((single) => {
			if (single.id === topic) {
				const tempId = topic + (single.messages.length + 1)
				const newMessage = { id: tempId, message: message, topicID: topic }
				single.messages = [...single.messages, newMessage]
			}
		})

		setData(dataCopy)
	}

	return (
		<div className="flex justify-center">
			<form
				autoComplete="off"
				onSubmit={handleSubmit}
				className="bg-white shadow-lg px-16 py-16 self-center my-4 lg:w-2/5 lg:max-w-md rounded-md w-full"
			>
				<h2 className="text-white text-4xl mb-8 font-semibold"></h2>
				<div className="flex flex-col gap-4">
					<TopicSelect
						data={data}
						value={topic}
						onChange={(e: any) => setTopic(e.target.value)}
						label="Topic"
						error={topicError}
					/>
					<TextField
						className="primary-textfield"
						label="Message"
						onChange={(e) => setMessage(e.target.value)}
						variant="outlined"
						type="text"
						sx={{ mb: 3 }}
						fullWidth
						value={message}
						error={messageError}
					/>
				</div>
				<Button className="primary-button w-full" type="submit">
					Add message
				</Button>
			</form>
		</div>
	)
}

export default Form
