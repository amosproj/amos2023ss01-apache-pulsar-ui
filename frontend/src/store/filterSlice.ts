// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2010-2021 Dirk Riehle <dirk@riehle.org
// SPDX-FileCopyrightText: 2019 Georg Schwarz <georg. schwarz@fau.de>

import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { RootState } from '.'

export type FilterState = {
	cluster: string[]
	tenant: string[]
	namespace: string[]
	topic: string[]
	message: string[]
}

type UpdateSingleFilter = {
	filterName: 'cluster' | 'tenant' | 'namespace' | 'topic' | 'message'
	id: string
}

const initialState: FilterState = {
	cluster: [],
	tenant: [],
	namespace: [],
	topic: [],
	message: [],
}

const filterSlice = createSlice({
	name: 'filterControl',
	initialState,
	reducers: {
		setCluster: (state, action: PayloadAction<string[]>) => {
			state.cluster = action.payload
		},
		setTenant: (state, action: PayloadAction<string[]>) => {
			state.tenant = action.payload
		},
		setNamespace: (state, action: PayloadAction<string[]>) => {
			state.namespace = action.payload
		},
		setTopic: (state, action: PayloadAction<string[]>) => {
			state.topic = action.payload
		},
		// Adds query to one single filter (cluster, tenant, namespace, topic)
		addFilter: (state, action: PayloadAction<UpdateSingleFilter>) => {
			const filterName = action.payload.filterName
			state[filterName].push(action.payload.id)
		},
		// Deletes query from one single filter (cluster, tenant, namespace, topic)
		deleteFilter: (state, action: PayloadAction<UpdateSingleFilter>) => {
			const filterName = action.payload.filterName
			const query = action.payload.id
			state[filterName] = state[filterName].filter((element) => {
				return element !== query
			})
		},
		resetAllFilters: (state) => {
			state = initialState
		},
	},
})

// Selectors
const selectCluster = (state: RootState): string[] => {
	return state.filterControl.cluster
}
const selectTenant = (state: RootState): string[] => {
	return state.filterControl.tenant
}
const selectNamespace = (state: RootState): string[] => {
	return state.filterControl.namespace
}
const selectTopic = (state: RootState): string[] => {
	return state.filterControl.topic
}
export { selectCluster, selectNamespace, selectTenant, selectTopic }

export const {
	setCluster,
	setTenant,
	setNamespace,
	setTopic,
	addFilter,
	deleteFilter,
	resetAllFilters,
} = filterSlice.actions

export default filterSlice.reducer
