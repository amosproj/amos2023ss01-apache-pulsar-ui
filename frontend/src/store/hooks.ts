// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2023 Julian Tochman-Szewc <tochman-szewc@campus.tu-berlin.de>
// SPDX-FileCopyrightText: 2023 Shahraz Nasir <shahraz.nasir@campus.tu-berlin.de>
// SPDX-FileCopyrightText: 2023 Ziqi He <ziqi.he@fau.de>
import { useDispatch, useSelector } from 'react-redux'
import type { TypedUseSelectorHook } from 'react-redux'
import type { RootState, AppDispatch } from './index'

// Use throughout app instead of plain `useDispatch` and `useSelector`
export const useAppDispatch: () => AppDispatch = useDispatch
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector
