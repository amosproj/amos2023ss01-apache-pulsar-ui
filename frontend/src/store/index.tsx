// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2010-2021 Dirk Riehle <dirk@riehle.org
// SPDX-FileCopyrightText: 2019 Georg Schwarz <georg. schwarz@fau.de>

import { configureStore } from '@reduxjs/toolkit'
import rootReducer from './reducers'

const store = configureStore({ reducer: rootReducer })

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch

export default store
