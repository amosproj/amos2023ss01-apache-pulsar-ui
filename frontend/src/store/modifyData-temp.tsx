interface Cluster {
	id: string
	tenants: Tenant[]
	brokers: string[]
	amountOfTenants: number
	amountOfNamespaces: number
	amountOfTopics: number
	amountOfBrokers: number
	brokerServiceUrl: string
	serviceUrl: string
}

interface Tenant {
	id: string
	namespaces: Namespace[]
	amountOfNamespaces: number
	amountOfTopics: number
	tenantInfo: TenantInfo
}

interface TenantInfo {
	adminRoles: string[]
	allowedClusters: string[]
}

interface Namespace {
	id: string
	topics: (string | Topic)[]
	amountOfTopics: number
	bundlesData: BundlesData
	messagesTTL: any // Use 'any' for now, you might want to replace it with the actual type
	retentionPolicies: any // Use 'any' for now, you might want to replace it with the actual type
}

interface BundlesData {
	boundaries: string[]
	numBundles: number
}

interface Topic {
	name: string
	localName: string
	namespace: string
	tenant: string
	ownerBroker: string
	topicStatsDto: {
		subscriptions: string[]
		producers: string[]
		numberSubscriptions: number
		numberProducers: number
		producedMesages: number
		consumedMessages: number
		averageMessageSize: number
		storageSize: number
	}
	producedMessages: number
	consumedMessages: number
	messagesDto: []
	persistent: boolean
}

interface TopicData {
	topics: Array<Topic>
}

/**
 * Takes in standard cluster data from sprint release 5 and converts it to dummy data structure
 * @param cluster
 * @returns modified data
 */
export const modifyData = (
	clusters: Array<Cluster>,
	topics: TopicData
): Array<Cluster> => {
	const newData = clusters.map((cluster: Cluster) => {
		const clusterCopy = { ...cluster }
		let newCluster = addClusterAndTenant(clusterCopy)
		newCluster = replaceTopicStrings(clusterCopy, topics)
		return newCluster
	})

	return newData
}

/**
 * Takes in standard cluster from sprint release 5 and converts it to dummy data structure
 * @param cluster
 * @returns modified cluster
 */
const addClusterAndTenant = (cluster: Cluster): Cluster => {
	// Add 'cluster' property to tenants
	cluster.tenants = cluster.tenants.map((tenant) => {
		// Add 'cluster' property to tenant
		const tenantWithCluster = { ...tenant, cluster: cluster.id }

		// Add 'cluster' and 'tenant' property to namespaces
		tenantWithCluster.namespaces = tenant.namespaces.map((namespace) => ({
			...namespace,
			cluster: cluster.id,
			tenant: tenant.id,
		}))

		return tenantWithCluster
	})

	return cluster
}

const replaceTopicStrings = (cluster: Cluster, topics: TopicData): Cluster => {
	const topicsMap = new Map(topics.topics.map((topic) => [topic.name, topic]))

	// Replace topic strings with topic objects. non strings stay the same
	cluster.tenants.forEach((tenant) => {
		tenant.namespaces.forEach((namespace) => {
			namespace.topics = namespace.topics.map((topicName) => {
				if (typeof topicName !== 'string') return topicName
				const topic = topicsMap.get(topicName)
				if (topic) return topic
				else return topicName
			})
			const filteredTopics: string[] = []
			namespace.topics = namespace.topics.filter((topic) => {
				const pass = typeof topic !== 'string'
				if (!pass) {
					filteredTopics.push(topic)
				}
				return pass
			})
			if (!filteredTopics) console.log('Filtered out topics: ', filteredTopics)
		})
	})

	return cluster
}
