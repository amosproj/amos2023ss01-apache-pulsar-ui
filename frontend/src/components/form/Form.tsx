import React, { useState, useEffect } from 'react'
import { TextField, Button } from '@mui/material'
import TopicSelect from './CustomSelect'
import CustomAccordion from './CustomAccordion'
import { SelectChangeEvent } from '@mui/material/Select'
import {
	selectData,
	selectMessage,
	selectTopic,
	setData,
	setMessage,
	setTopic,
	updateData,
} from './formControllerSlice'
import { useAppDispatch, useAppSelector } from '../../store/hooks'

const Form = () => {
	const [topicError, setTopicError] = useState<boolean>(false)
	const [messageError, setMessageError] = useState<boolean>(false)

	const data = useAppSelector(selectData)
	const topic = useAppSelector(selectTopic)
	const message = useAppSelector(selectMessage)

	const dispatch = useAppDispatch()

	//can later on be replaced by the fetchDataThunk
	const getData = () => {
		fetch('dummy/dummy.json', {
			headers: {
				'Content-Type': 'application/json',
				Accept: 'application/json',
			},
		})
			.then(function (response) {
				console.log(response)
				return response.json()
			})
			.then(function (json) {
				console.log(json)
				dispatch(setData(json))
			})
	}
	useEffect(() => {
		getData()
	}, [])

	const handleSubmit = (event: { preventDefault: () => void }) => {
		event.preventDefault()

		setMessageError(false)
		setTopicError(false)

		if (message == '') {
			setMessageError(true)
		}
		if (topic == '') {
			setTopicError(true)
		}

		if (message && topic) {
			console.log(message, topic)
		}

		dispatch(updateData({ message: message, topic: topic }))
	}

	return (
		<div className="flex justify-center gap-8 align-start">
			<form
				data-testid="demo-message-form"
				autoComplete="off"
				onSubmit={handleSubmit}
				className="bg-white shadow-lg px-16 py-16 self-center my-4 lg:w-2/5 lg:max-w-md rounded-md w-full"
			>
				<h4 className="text-black text-center text-3xl mb-4 font-semibold">
					Message Management
				</h4>
				<p className="text-center text-black pb-4">
					Use the form below to add a message to one of the existing Topics.
				</p>
				<div className="flex flex-col gap-4">
					<TopicSelect
						data={data}
						value={topic}
						onChange={(e: SelectChangeEvent<string>) =>
							dispatch(setTopic(e.target.value))
						}
						label="Topic"
						error={topicError}
					/>
					<TextField
						inputProps={{ 'data-testid': 'demo-simple-texfield' }}
						className="primary-textfield"
						label="Message"
						onChange={(e: React.ChangeEvent<HTMLInputElement>) =>
							dispatch(setMessage(e.target.value))
						}
						variant="outlined"
						type="text"
						sx={{ mb: 3 }}
						fullWidth
						value={message}
						error={messageError}
					/>
				</div>
				<Button className="primary-button w-full" type="submit">
					Add message
				</Button>
			</form>
			<div className="bg-white shadow-lg px-16 py-16 self-center my-4 lg:w-2/5 lg:max-w-md rounded-md w-full">
				<h4 className="text-black text-center text-3xl my-4 font-semibold">
					Topic View
				</h4>
				<p className="text-center text-black pb-4">
					Click on a Topic to view the associated Messages.
				</p>
				<CustomAccordion data={data} />
			</div>
		</div>
	)
}

export default Form
