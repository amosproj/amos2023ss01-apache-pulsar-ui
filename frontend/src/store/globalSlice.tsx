// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2010-2021 Dirk Riehle <dirk@riehle.org
// SPDX-FileCopyrightText: 2019 Georg Schwarz <georg. schwarz@fau.de>

import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit'
import axios from 'axios'
import { RootState } from '.'
import { modifyData } from './modifyData-temp'
import { Cluster } from 'cluster'

export type View = {
	selectedNav: string | null
	filteredId: number | string | null
}

export type globalState = {
	showLP: boolean
	view: View
	data: Array<MessageList>
	endpoint: string
	rawClusterData: any
	rawTopicData: any
	clusterData: any
	messageList: any[]
}

const initialState: globalState = {
	showLP: true,
	view: {
		selectedNav: 'cluster',
		filteredId: null,
	},
	data: [],
	endpoint: '127.0.0.1:8080',
	rawClusterData: [],
	rawTopicData: {},
	clusterData: [],
	messageList: [],
}

const backendInstance = axios.create({
	baseURL: 'http://localhost:8081/api',
	timeout: 1000,
})

const fetchRawClusterDataThunk = createAsyncThunk(
	'globalController/fetchData',
	async () => {
		const response = await backendInstance.get('/cluster')
		return response.data
	}
)

const fetchRawTopicDataThunk = createAsyncThunk(
	'globalController/fetchTopic',
	async () => {
		const response = await backendInstance.get('/topic/all')
		return response.data
	}
)

const fetchMessageDataThunk = createAsyncThunk(
	'globalController/fetchMessage',
	async ({ topic, subscription }: { topic: string; subscription: string }) => {
		const response = await backendInstance.get('/messages', {
			params: { topic: topic, subscription: subscription },
		})
		return response.data.messages
	}
)

const fetchAllMessagesThunk = createAsyncThunk(
	'globalController/fetchAllMessages',
	async (_, thunkAPI) => {
		const { dispatch } = thunkAPI
		const state = thunkAPI.getState() as RootState
		const promises: Promise<any>[] = []
		const { clusterData } = state.globalControl
		// iterates through every topic-subscription pair and dispatches a request for it
		clusterData.forEach((cluster: SampleCluster) => {
			cluster.tenants.forEach((tenant: SampleTenant) => {
				tenant.namespaces.forEach((namespace: SampleNamespace) => {
					namespace.topics.forEach((topic: SampleTopic) => {
						topic.topicStatsDto.subscriptions.forEach((sub: string) => {
							if (sub)
								promises.push(
									dispatch(
										fetchMessageDataThunk({
											topic: topic.name,
											subscription: sub,
										})
									)
								)
						})
					})
				})
			})
		})
		//awaits all requests
		const messagesResults = await Promise.all(promises)

		//combines all messages in an array
		const messages = [].concat(
			...messagesResults.map((result) => result.payload)
		)
		return messages
	}
)

const combineAsyncThunk = createAsyncThunk(
	'globalController/combineData',
	async (_, thunkAPI) => {
		const { dispatch } = thunkAPI
		// dispatch both thunks and wait for them to complete
		await Promise.all([
			dispatch(fetchRawClusterDataThunk()),
			dispatch(fetchRawTopicDataThunk()),
		])
	}
)

// eslint-disable-next-line
const globalSlice = createSlice({
	name: 'globalControl',
	initialState,
	reducers: {
		/** Old landing page logic */
		// moveToApp: (state) => {
		// 	state.showLP = false
		// 	console.log('sdsd')
		// },
		// backToLP: () => initialState,
		/** End of landing page logic */
		setNav: (state, action: PayloadAction<string>) => {
			state.view.selectedNav = action.payload
		},
		setView: (state, action: PayloadAction<View>) => {
			state.view = action.payload
		},
		setData: (
			state: globalState,
			action: PayloadAction<Array<MessageList>>
		) => {
			state.data = action.payload
		},
		updateData: (state: globalState, action: PayloadAction<UpdateForData>) => {
			state.data.map((single: MessageList) => {
				if (single.id === action.payload.topic) {
					const tempId = action.payload.topic + (single.messages.length + 1)
					const newMessage = {
						id: tempId,
						value: action.payload.message,
						topic: action.payload.topic,
					}
					single.messages = [...single.messages, newMessage]
				}
				return single
			})
		},
		setEndpoint: (state: globalState, action: PayloadAction<string>) => {
			state.endpoint = action.payload
		},
	},
	extraReducers: (builder) => {
		builder.addCase(fetchRawClusterDataThunk.fulfilled, (state, action) => {
			state.rawClusterData = JSON.parse(JSON.stringify(action.payload))
		})
		builder.addCase(fetchRawClusterDataThunk.rejected, (state) => {
			console.log('fetch raw cluster data failed')
		})
		builder.addCase(fetchRawTopicDataThunk.fulfilled, (state, action) => {
			state.rawTopicData = JSON.parse(JSON.stringify(action.payload))
		})
		builder.addCase(fetchRawTopicDataThunk.rejected, (state) => {
			console.log('fetch raw topic data failed')
		})
		builder.addCase(combineAsyncThunk.fulfilled, (state, action) => {
			//combine raw data to meet dummy data structure and save it in clusterData
			state.clusterData = modifyData(state.rawClusterData, state.rawTopicData)
		})
		builder.addCase(combineAsyncThunk.rejected, (state, action) => {
			console.log('combine Async thunk failed')
			state.clusterData = []
			state.rawClusterData = []
			state.rawTopicData = {}
		})
		builder.addCase(fetchAllMessagesThunk.fulfilled, (state, action) => {
			state.messageList = action.payload
		})
		builder.addCase(fetchAllMessagesThunk.rejected, (state) => {
			console.log('fetch all messages thunk failed')
		})
	},
})

const { actions, reducer } = globalSlice

const selectData = (state: RootState): Array<MessageList> =>
	state.globalControl.data

const selectView = (state: RootState): View => state.globalControl.view

const selectShowLP = (state: RootState): boolean => state.globalControl.showLP

const selectEndpoint = (state: RootState): string =>
	state.globalControl.endpoint

const selectClusterData = (state: RootState): any =>
	state.globalControl.clusterData

const selectMessages = (state: RootState): any =>
	state.globalControl.messageList

export const { setNav, setView, updateData, setData, setEndpoint } = actions

export {
	selectShowLP,
	selectEndpoint,
	selectData,
	selectView,
	selectClusterData,
	combineAsyncThunk,
	selectMessages,
	fetchAllMessagesThunk,
}

export default reducer
