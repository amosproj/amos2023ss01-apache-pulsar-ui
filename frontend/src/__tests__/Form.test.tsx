import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import Form from '../components/form/Form'
import { Provider } from 'react-redux'
import store from '../store'

test('should add element by submitting form', async () => {
	const formComponent = render(
		<Provider store={store}>
			<Form />
		</Provider>
	)
	const selectinput = formComponent.getByTestId('demo-simple-select')
	fireEvent.change(selectinput, { target: { value: '100' } })
	const textinput = formComponent.getByTestId('demo-simple-texfield')
	fireEvent.change(textinput as HTMLInputElement, {
		target: { value: 'Test with Jest' },
	})
	expect((textinput as HTMLInputElement).value).toBe('Test with Jest')
	const form = await formComponent.findByTestId('demo-message-form')
	fireEvent.submit(form)

	expect(formComponent).toMatchSnapshot()
})
