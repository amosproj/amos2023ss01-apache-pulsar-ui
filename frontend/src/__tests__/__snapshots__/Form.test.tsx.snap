// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add element by submitting form 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex justify-center gap-8 align-start"
      >
        <form
          autocomplete="off"
          class="bg-white shadow-lg px-16 py-16 self-center my-4 lg:w-2/5 lg:max-w-md rounded-md w-full"
          data-testid="demo-message-form"
        >
          <h4
            class="text-black text-center text-3xl mb-4 font-semibold"
          >
            Message Management
          </h4>
          <p
            class="text-center text-black pb-4"
          >
            Use the form below to add a message to one of the existing Topics.
          </p>
          <div
            class="flex flex-col gap-4"
          >
            <div
              class="MuiBox-root css-ece9u5"
            >
              <div
                class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                  id="demo-simple-select-label"
                >
                  Topic
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-error MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
                >
                  <div
                    aria-expanded="false"
                    aria-haspopup="listbox"
                    aria-labelledby="demo-simple-select-label"
                    class="MuiSelect-select MuiSelect-outlined Mui-error MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                    role="button"
                    tabindex="0"
                  >
                    <span
                      class="notranslate"
                    >
                      ​
                    </span>
                  </div>
                  <input
                    aria-hidden="true"
                    aria-invalid="true"
                    class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                    data-testid="demo-simple-select"
                    tabindex="-1"
                    value=""
                  />
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                    data-testid="ArrowDropDownIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-yjsfm1"
                    >
                      <span>
                        Topic
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root primary-textfield css-32yp37-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="true"
                for=":r0:"
                id=":r0:-label"
              >
                Message
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  data-testid="demo-simple-texfield"
                  id=":r0:"
                  type="text"
                  value="Test with Jest"
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-14lo706"
                  >
                    <span>
                      Message
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium primary-button w-full css-1e6y48t-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="submit"
          >
            Add message
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </form>
        <div
          class="bg-white shadow-lg px-16 py-16 self-center my-4 lg:w-2/5 lg:max-w-md rounded-md w-full"
        >
          <h4
            class="text-black text-center text-3xl my-4 font-semibold"
          >
            Topic View
          </h4>
          <p
            class="text-center text-black pb-4"
          >
            Click on a Topic to view the associated Messages.
          </p>
          <div
            data-testid="demo-accordion"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex justify-center gap-8 align-start"
    >
      <form
        autocomplete="off"
        class="bg-white shadow-lg px-16 py-16 self-center my-4 lg:w-2/5 lg:max-w-md rounded-md w-full"
        data-testid="demo-message-form"
      >
        <h4
          class="text-black text-center text-3xl mb-4 font-semibold"
        >
          Message Management
        </h4>
        <p
          class="text-center text-black pb-4"
        >
          Use the form below to add a message to one of the existing Topics.
        </p>
        <div
          class="flex flex-col gap-4"
        >
          <div
            class="MuiBox-root css-ece9u5"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                id="demo-simple-select-label"
              >
                Topic
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-error MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
              >
                <div
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-labelledby="demo-simple-select-label"
                  class="MuiSelect-select MuiSelect-outlined Mui-error MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                  role="button"
                  tabindex="0"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </div>
                <input
                  aria-hidden="true"
                  aria-invalid="true"
                  class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                  data-testid="demo-simple-select"
                  tabindex="-1"
                  value=""
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      Topic
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root primary-textfield css-32yp37-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              for=":r0:"
              id=":r0:-label"
            >
              Message
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                data-testid="demo-simple-texfield"
                id=":r0:"
                type="text"
                value="Test with Jest"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-14lo706"
                >
                  <span>
                    Message
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium primary-button w-full css-1e6y48t-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="submit"
        >
          Add message
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </form>
      <div
        class="bg-white shadow-lg px-16 py-16 self-center my-4 lg:w-2/5 lg:max-w-md rounded-md w-full"
      >
        <h4
          class="text-black text-center text-3xl my-4 font-semibold"
        >
          Topic View
        </h4>
        <p
          class="text-center text-black pb-4"
        >
          Click on a Topic to view the associated Messages.
        </p>
        <div
          data-testid="demo-accordion"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
